### ISSUE TITLE
Poor IPsec performance on Kubernetes

### ISSUE THREAD

**Issue by amukherj** [cilium/cilium#26648]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

1. Kubernetes version 1.27.2 on KVM VMs with Open VSwitch-based virtual networking. Cilium 1.13.0 / 1.13.2 for cluster networkng. MTU is 65000 on the virtual NICs all the way down to the OVS bridge on the host.
2. 3-node cluster with two workers. Each VM has 8 GiB of RAM, and 6 vcpu cores (3 vcpus x 2 cores). Cilium transparent encryption enabled using the following helm parameters:
        encryption.type="ipsec"
        encryption.enabled=true
        encryption.ipsec.interface="eth0"
        encryption.ipsec.keyFile="keys"
        encryption.ipsec.mountPath="/etc/ipsec/keys"
        encryption.ipsec.secretName="cilium-ipsec-keys"
3. Ran iperf3 server in a pod on one worker, and iperf3 client (single threaded) from a pod on the second worker to connect to the first. The following are the iperf3 command-line details.
        iperf3 -s -p <port>           # in server pod on worker1
        iperf3 -c <server_pod_ip> -p <port> -t 60   # in client pod on worker2
4. When transparent encryption is not enabled, I see a throughput of around 2.7 Gbps. Once I enable transparent encryption, the throughput drops to 200 Mbps.

### Cilium Version

1.13.0 / 1.13.2

### Kernel Version

6.3.3 (CentOS Stream 9, kernel version 6.3.3-1.el7.elrepo.x86_64)

### Kubernetes Version

1.27.2

### Sysdump

[cilium-sysdump-20230705-141625.zip](https://github.com/cilium/cilium/files/11958881/cilium-sysdump-20230705-141625.zip)

### Anything else?

Here are individual flame graphs of the client and server side CPUs.

#### Client flame graphs
![client_cpu_0](https://github.com/cilium/cilium/assets/2528112/f8016e19-4425-470c-a3db-8d4c6f147612)
![client_cpu_1](https://github.com/cilium/cilium/assets/2528112/0465c111-6a03-49a6-8be5-eab9d102759d)
![client_cpu_2](https://github.com/cilium/cilium/assets/2528112/71fc176e-e59b-480e-b3fe-b48c496f4b88)
![client_cpu_3](https://github.com/cilium/cilium/assets/2528112/abad930f-d24b-4453-aabe-de851c906c38)
![client_cpu_4](https://github.com/cilium/cilium/assets/2528112/ca91adc4-53be-4475-a740-bd929595bea3)
![client_cpu_5](https://github.com/cilium/cilium/assets/2528112/2ba88f3d-a9a7-4139-be29-7a0129893de2)

#### Server flame graphs
![server_cpu_0](https://github.com/cilium/cilium/assets/2528112/508da255-8fe4-4791-abd2-59c2dcddd8ea)
![server_cpu_1](https://github.com/cilium/cilium/assets/2528112/040a9fe9-7652-4474-9270-9b05f9577351)
![server_cpu_2](https://github.com/cilium/cilium/assets/2528112/38f12754-61f2-40b7-80da-292ed7427393)
![server_cpu_3](https://github.com/cilium/cilium/assets/2528112/f5afe898-a128-416e-814e-6d8627e6b874)
![server_cpu_4](https://github.com/cilium/cilium/assets/2528112/2bb78a8d-d440-4b1e-b5b7-7d2e91b87cc6)
![server_cpu_5](https://github.com/cilium/cilium/assets/2528112/d3d3b19d-29bc-47e0-afbb-abd12e53b317)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by learnitall** [comment]
Taking a look, there are two things that stand out: 
1. All the graphs except two show idle CPU time as the majority. This is expected for IPSec. It is difficult to saturate more than one CPU with IPSec traffic, as the kernel's RSS mechanism will tunnel all traffic onto a single core (at least on the receiving side).
2. Large stacks starting with `tcf_classify`, with the majority of the stack taken up by `bpf_skb_event_output`. Looks like there are a couple of different programs which are sending lots of data in the perf ring buffer up to userspace, which is taking up a lot of CPU. I dug around and there are two areas where we send trace events on every IPSec packet (in your environment). 
  * [`cil_from_overlay`](https://github.com/cilium/cilium/blob/main/bpf/bpf_overlay.c#L555) -> [`send_trace_notify`](https://github.com/cilium/cilium/blob/main/bpf/lib/trace.h#L185)
  * [`tail_handle_ipv4`](https://github.com/cilium/cilium/blob/main/bpf/bpf_overlay.c#L442) -> [`handle_ipv4`](https://github.com/cilium/cilium/blob/main/bpf/bpf_overlay.c#L287) -> [`send_trace_notify`](https://github.com/cilium/cilium/blob/main/bpf/lib/trace.h#L185) (called in `handle_ipv4` [here](https://github.com/cilium/cilium/blob/main/bpf/bpf_overlay.c#L410)

I'll dig into this more as time is available and talk with some other folks to gather more information about what is going on here.

**Comment by pchaigno** [comment]
Nice investigation @learnitall!

I believe we should switch `TRACE_PAYLOAD_LEN` to `0` in https://github.com/cilium/cilium/blob/755eae8377fc4c353337af31be5d502c333b9f3c/bpf/bpf_overlay.c#L410-L412
according to https://github.com/cilium/cilium/blob/755eae8377fc4c353337af31be5d502c333b9f3c/bpf/lib/trace.h#L172-L178

It's probably not the only place where we incorrectly set `TRACE_PAYLOAD_LEN`.

**Comment by pchaigno** [comment]
Oh, nice! I didn't expect it would make a big difference.

**Comment by learnitall** [comment]
I went through and found two other places where I think this change is appropriate and then finished up some testing today. I will post more details tomorrow with a PR, but here is a quick before/after on a netperf TCP stream test on Cilium 1.14:

Before:

(Each row is a separate netperf client)

```csv
Throughput,Throughput Units,Elapsed Time (sec)
464.06,10^6bits/s,180.04
461.25,10^6bits/s,180.02
441.73,10^6bits/s,180.00
482.26,10^6bits/s,180.04
```

Server Node:

![Image](https://github.com/cilium/cilium/assets/13142290/d344eea2-a24b-40dd-a42d-4b0b07e44a24)

Client Node:

![Image](https://github.com/cilium/cilium/assets/13142290/e87beb6b-6605-4874-89c0-c65bddd9e726)


After:

```csv
Throughput,Throughput Units,Elapsed Time (sec)
698.48,10^6bits/s,180.03
643.81,10^6bits/s,180.05
780.22,10^6bits/s,180.05
552.11,10^6bits/s,180.03
```

Server node:

![Image](https://github.com/cilium/cilium/assets/13142290/3ea878d0-3960-4f12-b290-b4254acb5538)

Client node:

![Image](https://github.com/cilium/cilium/assets/13142290/6c81b0bb-2e3a-4185-b0ca-d041c0b5aaa3)

Averaging these runs and comparing the two, there's a 45% increase in throughput with the change.

Patch:

```diff
diff --git a/bpf/bpf_overlay.c b/bpf/bpf_overlay.c
index f7a1880ea8..566b79e8e8 100644
--- a/bpf/bpf_overlay.c
+++ b/bpf/bpf_overlay.c
@@ -123,8 +123,7 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
                ctx_change_type(ctx, PACKET_HOST);
 
                send_trace_notify(ctx, TRACE_TO_STACK, *identity, 0, 0,
-                                 ctx->ingress_ifindex, TRACE_REASON_ENCRYPTED,
-                                 TRACE_PAYLOAD_LEN);
+                                 ctx->ingress_ifindex, TRACE_REASON_ENCRYPTED, 0);
 
                return CTX_ACT_OK;
        }
@@ -408,8 +407,7 @@ skip_vtep:
                ctx_change_type(ctx, PACKET_HOST);
 
                send_trace_notify(ctx, TRACE_TO_STACK, *identity, 0, 0,
-                                 ctx->ingress_ifindex, TRACE_REASON_ENCRYPTED,
-                                 TRACE_PAYLOAD_LEN);
+                                 ctx->ingress_ifindex, TRACE_REASON_ENCRYPTED, 0);
 
                return CTX_ACT_OK;
        }
@@ -588,8 +586,7 @@ int cil_from_overlay(struct __ctx_buff *ctx)
 #ifdef ENABLE_IPSEC
        if (is_esp(ctx, proto))
                send_trace_notify(ctx, TRACE_FROM_OVERLAY, 0, 0, 0,
-                                 ctx->ingress_ifindex, TRACE_REASON_ENCRYPTED,
-                                 TRACE_PAYLOAD_LEN);
+                                 ctx->ingress_ifindex, TRACE_REASON_ENCRYPTED, 0);
        else
 #endif
        {
```

**Comment by tejas-sangol** [comment]
@learnitall Thanks for fixing this! Is there any way we can install the latest commits using Helm/CLI?

**Comment by amukherj** [comment]
@learnitall preliminary findings with the latest commit suggest that there is a definite improvement in throughput. The drop in throughput is no longer as precipitous - I see a 70% drop in throughput with IPsec on, as opposed to 92.5% earlier. I don't know what percentage drop should be considered par. Will report more findings as we have it. cc: @tejas-sangol

**Comment by learnitall** [comment]
That's great @amukherj, glad to hear there is a bit of an improvement. @tejas-sangol the commit was released as part of 1.14.1 last week

**Comment by pchaigno** [comment]
@amukherj If you have updated flamegraphs, we can take a look to see if there are any other low-hanging fruits. That said, this is likely the consequence of the RSS limitation (last limitation at https://docs.cilium.io/en/stable/security/network/encryption-ipsec/#limitations). Removing that RSS limitation is a larger work item which hasn't been prioritized so far.
